pkgs/top-level/haskell-packages.nix

```diff
@@ -115,68 +118,78 @@ in rec {
   # Default overrides that are applied to all package sets.
   packageOverrides = self : super : {};
 
-  # Always get compilers from `buildPackages`
-  packages = let bh = buildPackages.haskell; in {
+  packages = {
 
     ghc7103 = callPackage ../development/haskell-modules {
+      buildCompiler = bh.compiler.ghc7103;
       buildHaskellPackages = bh.packages.ghc7103;
-      ghc = bh.compiler.ghc7103;
+      compiler = compiler.ghc7103;
       compilerConfig = callPackage ../development/haskell-modules/configuration-ghc-7.10.x.nix { };
     };
     ghc7103Binary = callPackage ../development/haskell-modules {
+      buildCompiler = bh.compiler.ghc7103Binary;
       buildHaskellPackages = bh.packages.ghc7103Binary;
-      ghc = bh.compiler.ghc7103Binary;
+      compiler = compiler.ghc7103Binary;
       compilerConfig = callPackage ../development/haskell-modules/configuration-ghc-7.10.x.nix { };
       packageSetConfig = bootstrapPackageSet;
     };
     ghc802 = callPackage ../development/haskell-modules {
+      buildCompiler = bh.compiler.ghc802;
       buildHaskellPackages = bh.packages.ghc802;
-      ghc = bh.compiler.ghc802;
+      compiler = compiler.ghc802;
       compilerConfig = callPackage ../development/haskell-modules/configuration-ghc-8.0.x.nix { };
     };
     ghc821Binary = callPackage ../development/haskell-modules {
+      buildCompiler = bh.compiler.ghc821Binary;
       buildHaskellPackages = bh.packages.ghc821Binary;
-      ghc = bh.compiler.ghc821Binary;
-      compilerConfig = callPackage ../development/haskell-modules/configuration-ghc-8.2.x.nix { };
+      compiler = compiler.ghc821Binary;
+      compilerConfig = callPackage ../development/haskell-modules/configuration-ghc-8.2.x-binary.nix { };
       packageSetConfig = bootstrapPackageSet;
     };
     ghc822 = callPackage ../development/haskell-modules {
+      buildCompiler = bh.compiler.ghc822;
       buildHaskellPackages = bh.packages.ghc822;
-      ghc = bh.compiler.ghc822;
+      compiler = compiler.ghc822;
       compilerConfig = callPackage ../development/haskell-modules/configuration-ghc-8.2.x.nix { };
     };
     ghc843 = callPackage ../development/haskell-modules {
+      buildCompiler = bh.compiler.ghc843;
       buildHaskellPackages = bh.packages.ghc843;
-      ghc = bh.compiler.ghc843;
+      compiler = compiler.ghc843;
       compilerConfig = callPackage ../development/haskell-modules/configuration-ghc-8.4.x.nix { };
     };
     ghcHEAD = callPackage ../development/haskell-modules {
+      buildCompiler = bh.compiler.ghcHEAD;
       buildHaskellPackages = bh.packages.ghcHEAD;
-      ghc = bh.compiler.ghcHEAD;
+      compiler = compiler.ghcHEAD;
       compilerConfig = callPackage ../development/haskell-modules/configuration-ghc-head.nix { };
     };
     ghcjs = packages.ghcjs82;
     ghcjs710 = callPackage ../development/haskell-modules rec {
-      buildHaskellPackages = ghc.bootPkgs;
```

Comment:
I have no idea how ghcjs works, but I really don’t understand this part. `buildHaskellPackages` are packages used for `Setup.hs` (at least I think I read something like this in comments). So why is it `ghcBootPkgs` here?
(I guess, cc @Ericson2314 as well)